### PR TITLE
Fix save-analysis generation crash with invalid tuple access

### PR DESCRIPTION
**Pull request by sinkuu**
Reproduction:

```rust
fn invalid_tuple_struct_accessing() {
    bar.0;
}
```
```
error[E0425]: cannot find value `bar` in this scope
 --> test.rs:2:5
  |
2 |     bar.0;
  |     ^^^ not found in this scope

error[E0601]: main function not found

error: internal compiler error: librustc_save_analysis/dump_visitor.rs:1678: Expected struct or tuple type, found TyError
 --> test.rs:2:5
  |
2 |     bar.0;
  |     ^^^^^

thread 'rustc' panicked at 'Box<Any>', librustc_errors/lib.rs:482:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

This should fix a crash in RLS when editing such code. cc @nrc

### CODE REVIEW

File: src/librustc_save_analysis/dump_visitor.rs [hunk 1]
```diff
@@ -1675,7 +1675,10 @@ impl<'l, 'tcx: 'l, 'll, O: DumpOutput + 'll> Visitor<'l> for DumpVisitor<'l, 'tc
                         }
                     }
                     ty::TyTuple(..) => {}
-                    _ => span_bug!(ex.span, "Expected struct or tuple type, found {:?}", ty),
+                    _ => {
```
Comment (nrc) [inline src/librustc_save_analysis/dump_visitor.rs]: When you're using the invalid tuple constructor is this ty:err or something similarly specific? Could we explicitly catch that case and return, and still `span_bug` ICE if it's something else? If we silently fail and cope then it's really hard to notice the omission until somebody reports a hard to discover bug against the RLS.

Comment (sinkuu) [reply src/librustc_save_analysis/dump_visitor.rs]: It seems that almost all type variants (all but `TyInfer`?) can be crafted.

```rust
fn foo<T>(x: T) {
    bar.0; // TyError
    (0).0; // TyError
    x.0; // TyParam
    static S: i32 = 0;
    (S).0; // TyInt
    [].0; // TyArray
    extern { type Externed; }
    (|| -> Externed { panic!() })().0; // TyForeign
    (|| -> ! { panic!() })().0; // TyNever
    ...
}
```

**Comment by nrc** [comment]
OK, makes sense. Second commit looks fine.

@bors: r+

**Pull request merged**